service/matching/task_queue_manager.go

```diff
@@ -809,20 +818,52 @@ func (c *taskQueueManagerImpl) RedirectToVersionedQueueForAdd(ctx context.Contex
 
 	if c.kind == enumspb.TASK_QUEUE_KIND_STICKY {
 		// In the sticky case we don't redirect, but we may kick off this worker if there's a newer one.
-		err := checkVersionForStickyAdd(data, buildId)
-		return c.taskQueueID, userDataChanged, err
+		unknownBuild, err := checkVersionForStickyAdd(data, buildId)
+		if err != nil {
+			return nil, nil, err
+		}
+		if unknownBuild {
+			c.recordUnknownBuildTask(buildId)
+			// Don't bother persisting the unknown build id in this case: sticky tasks have a
+			// short timeout, so it doesn't matter if they get lost.
```

Comment:
Why does sticky tasks having a short timeout imply that it doesn't matter if they're dropped?

Reply:
Because history will retry on the normal queue is 5 seconds anyway.

The reason we're persisting is because of this unlikely sequence:
1. Matching gets task with unknown build id; it guesses a version set id for it
2. There's no poller, so it writes it to the db
3. That task queue gets unloaded due to idleness
4. The replication finishes and it turns out that build id actually belonged to another set
5. Now a poller comes for it

If we don't persist the build id, the specific versioned task queue will be lost to that poller, since any new task or poll will find the other set for it. So for tasks with no timeout, that we're not allowed to drop, we need to persist so we can find it again. But for sticky tasks with a 5s timeout, the chances that all the above happens within 5s is extraordinarily unlikely, and even if it did happen, dropping the task is totally fine and easier